Assets/Scripts/Algorithms/Patrolling/CognitiveCoordinated/CognitiveCoordinatedVirtualStigmergy.cs

```diff
@@ -0,0 +1,60 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using Maes.Map;
+
+namespace Maes.Algorithms.Patrolling
+{
+    /// <summary>
+    /// This is a distributed version of the Cognitive Coordinated Algorithm.
+    /// It uses virtual stigmergy to coordinate the robots.
+    /// Each robot has its own local map and uses the occupied vertices of other robots to coordinate.
+    /// </summary>
+    public sealed class CognitiveCoordinatedVirtualStigmergy : CognitiveCoordinatedBase
+    {
+        public override string AlgorithmName => "Cognitive Coordinated (virtual-stigmergy knowledge) Algorithm";
+        protected override PatrollingMap _globalMap => _patrollingMap;
+
+        public override void InitializeCoordinator(PatrollingMap _)
+        {
+        }
+
+        public override void OccupyVertex(int robotId, Vertex vertex)
+        {
+#if DEBUG
+            if (!_globalMap.Vertices.Contains(vertex))
+            {
+                throw new ArgumentException($"Vertex ({vertex}) is not a part of GlobalMap.Vertices.", nameof(vertex));
+            }
+#endif
+            if (!_virtualStigmergyComponent.Has(VirtualStigmergyMessageId))
+            {
+                _virtualStigmergyComponent.Put(VirtualStigmergyMessageId, new Dictionary<int, Vertex>());
+            }
+
+            // Update the local knowledge of the robot.
+            var _localKnowledge = _virtualStigmergyComponent.GetNonSending(VirtualStigmergyMessageId)!;
+            _localKnowledge[robotId] = vertex;
```

Comment:
This is verboden, you should clone the dictionary and put the cloned dictionary, to make sure you are not smuggling information. Why are you not just using the virtual stigmergy as a dictionary? key: int (robot id), value int (vertex id).

Reply:
I am now, I was worrying about the amount of msgs

Reply:
don't worry about it. we can change the virtual stigmergy code to batch messages created on the same tick together into one.